### PR TITLE
Feature/p2363

### PR DESCRIPTION
**Pull request by cmazakas**
Implement the rest of the transparent operations as specified by: http://wg21.link/P2363

Addresses https://github.com/boostorg/unordered/issues/150

### CODE REVIEW

**Review by joaquintides — CHANGES_REQUESTED** [review]
* Additional to the in-code comments, the PR lacks documentation. I can do this over the weekend.

File: include/boost/unordered/detail/type_traits.hpp [hunk 1]
```diff
@@ -53,10 +53,10 @@ namespace boost {
       {
       };
 
-      template <class, class A, class B> struct are_transparent
+      template <class, class Hash, class KeyEqual> struct are_transparent
       {
         static bool const value =
-          is_transparent<A>::value && is_transparent<B>::value;
+          is_transparent<Hash>::value && is_transparent<KeyEqual>::value;
```
[the remark this answers is not in the document]
Comment (cmazakas) [reply include/boost/unordered/detail/type_traits.hpp:59]: The SFINAE for the transparent member functions only works when they're dependent on the template parameter so we use a dummy here to keep the check dependent.

File: include/boost/unordered/detail/implementation.hpp [hunk 2]
```diff
@@ -1477,6 +1477,33 @@ namespace boost {
...
+          return a.release();
+        }
+
+        template <typename T, typename V, typename Alloc, typename Key>
+        inline typename boost::allocator_pointer<Alloc>::type
+        try_construct_node(
+          std::pair<T const, V>*, Alloc& alloc, BOOST_FWD_REF(Key) k)
+        {
+          return construct_node_pair(alloc, boost::forward<Key>(k));
+        }
```
[the remark this answers is not in the document]
Comment (cmazakas) [reply include/boost/unordered/detail/implementation.hpp:1506]: Yeah, the `try_` prefix came from the fact that these functions are _only_ being used in `try_emplace_unique`.

Good call on the first overload not delegating properly as it should.

`construct_node_or_node_pair` is fine. Maybe something like `construct_node_from_key_only` could also work.

**Comment by joaquintides** [comment]
While updating the docs, I noticed that `unordered_map::try_emplace` (unlike `emplace`) does not add overloads up to 10 arguments to emulate variadic args, despite what the reference says. Note related to PR, you may want to file an issue.

**Review by joaquintides — CHANGES_REQUESTED** [review]
There's a bug with the P2363 spec that shows in our implementation of it:
```cpp
struct transparent_hash
{
  using is_transparent=void;

  template<typename T>
  std::size_t operator()(const T& x)const
  {
    return boost::hash<T>()(x);
  }
};

using transparent_set=boost::unordered_flat_set<int,transparent_hash,std::equal_to<>>;
...
transparent_set ts1,ts2;
ts1.insert(ts2.begin(),ts2.end());
```
The call to `insert` is ambiguous (could be heterogeneous-lookup `insert` with a hint or range `insert`).

**Review by joaquintides — APPROVED** [review]

File: doc/unordered/unordered_flat_map.adoc [hunk 3]
```diff
@@ -938,17 +971,25 @@ If there is an element with key `k`, then it is updated by assigning `boost::for
...
+           std::forward_as_tuple(boost::forward<M>(obj)))
 ```
 
 [horizontal]
 Returns:;; The `bool` component of the return type is `true` if an insert took place. +
 +
 If an insert took place, then the iterator points to the newly inserted element. Otherwise, it points to the element with equivalent key.
 Throws:;; If an exception is thrown by an operation other than a call to `hasher` the function has no effect.
-Notes:;; Can invalidate iterators pointers and references, but only if the insert causes the load to be greater than the maximum load. +
+Notes:;; Can invalidate iterators pointers and references, but only if the insert causes the load to be greater than the maximum load.  +
```
Comment (pdimov) [inline doc/unordered/unordered_flat_map.adoc:990]: Is `:;;` correct here? Shouldn't it be `:::`?

File: doc/unordered/unordered_multimap.adoc [hunk 4]
```diff
@@ -1330,11 +1305,13 @@ Returns:;; The number of elements in bucket `n`.
 ==== bucket
 ```c++
 size_type bucket(const key_type& k) const;
+templat<class K> size_type bucket(const K& k) const;
```
Comment (pdimov) [inline doc/unordered/unordered_multimap.adoc:1308]: typo: `templat`

File: doc/unordered/unordered_flat_map.adoc (hunk 3 above)
Comment (joaquintides) [reply doc/unordered/unordered_flat_map.adoc:990]: This is what I found when I begin updating the docs (no ocurrence of `:::` overall), and it displays OK, but other than this I don't know. Maybe @cmazakas knows more about Asciidoc syntax to shed some light here.

File: doc/unordered/unordered_multimap.adoc (hunk 4 above)
Comment (joaquintides) [reply doc/unordered/unordered_multimap.adoc]: Fixed

File: doc/unordered/unordered_flat_map.adoc (hunk 3 above)
Comment (pdimov) [reply doc/unordered/unordered_flat_map.adoc:990]: Maybe `;;` is an alternative token?

Comment (pdimov) [reply doc/unordered/unordered_flat_map.adoc:990]: Yes, it is. https://docs.asciidoctor.org/asciidoc/latest/lists/description/

**Review by pdimov — APPROVED** [review]

**Pull request merged**